hip-0000-smart-contract-hcs-interoperable.md

```diff
@@ -0,0 +1,58 @@
+---
+hip: 478
+title: Interoperability Between Smart Contracts and the HCS
+author: John Conway (@scalemaildev), Walter Hernandez (@walter-hernandez)
+type: Standards Track
+category: Core
+needs-council-approval: Yes
+status: Last Call
+last-call-date-time: 2022-06-21T07:00:00Z
+created: 05-17-2022
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/479
+updated: 05-24-2022
+---
+
+## Abstract
+
+Enable smart contract interaction with the Hedera Consensus Service, by allowing contracts to create, read, and write to HCS topics.
+
+## Motivation
+
+Interoperability between the HCS and smart contracts would enable developers to leverage HCS-stored data in their dApps.
+
+## Rationale
+
+1. By enabling smart contracts to interact with the HCS, this would allow the creation of HCS-based Oracles. For example, a smart contract could subscribe to a private or public topic in order to read its external data, such as the prices of crypto-assets.
+
+2. This would also enable topic-based NFT minting rights. A smart contract could ingest an HCS topic to calculate the user's available NFTs. Then, after minting the NFT and transferring it to the user, the smart contract could update that same HCS topic. Currently, this process is only achievable via a centralized, server-based chain of events (such as on [https://hashandslash.io](hashandslash.io)).
+
+3. Any other Oracle use case could be repurposed by the HCS.
+
+## User stories
+
+As a developer, I want to be able to read HCS topics from my smart contracts.
```

Comment:
@mgarbs Agree, all the data related part of oracles is a process that would be outside Hedera Smart Contracts Service because of the computing needed for data collection, cleaning, and calculation of most common values and/or median (in the case of numerical values) from the data collected. 

Now, the idea of reading HCS topics within Smart Contracts is to bring the data from outside Hedera's network into it in order that Hedera Smart Contract Service can expand the functionality that its contracts can have. For example, a function within a contract can get the latest cryptoasset prices from an Oracle Smart Contract and pays to this Oracle Smart Contract for it. The Oracle smart contract gets the latest cryptoasset prices by reading a private topic, which is updated from outside Hedera network

Let me know what you think and if what you see it feasible and/or if there are better approaches

Reply:
@walter-hernandez mgarbs is correct.
The smart contract at most would be able to read the topic entity information, submit a topic message but will not have access to individual topic messages already emitted.

Are you suggesting network changes to have access to past messages?

Reply:
@mgarbs thank you for your comments and suggestions about help to get something working! :-)

@Nana-EC In the case of past messages, an Oracle Smart Contract could be used for storing some of the topic messages that it has received from the topic and it could have logic to delete/update messages as needed

Reply:
@walter-hernandez given the complexity of what you're suggesting I recommend adding a sequence diagram or another appropriate architectural diagram showcasing your suggested flow.

I have a concept of what you're suggesting for the Oracle Smart Contract but don't want to assume as there seems to be a mixture of responsibilities between the a devs smart contract, the precompile smart contract on the EVM and the Hedera network.
Note that anything that impacts state on the EVM or network comes with performance concerns and higher gas costs (which is something you'll want to explore).

The message queue suggestion is an interesting one that comes up. However, I think that would be outside of this HIP and would likely require a HIP on its own if you're suggesting it be implemented at the network services level.

Reply:
@gregscullard HCS as a data feed and the consumption of its feed would be as a message queue system. Now, you raise a good consideration in the design about the frequency of messages and their consumption by HSCS.

In the case of a feed for an exchange rate, I guess the smart contract service could consume only the latest message in the queue and the oldest messages in the queue (HCS) should be removed. 

@gregscullard In general, how feasible do you think it is to implement some interoperability between HCS and smart contracts service? What do you think is pending in this PR for it to be closed and development to start?